### PR TITLE
egressip: add metrics

### PR DESCRIPTION
**Pull request by flavio-fernandes**
Add metrics to observe core events related to EgressIP:

- name: egress_ips_assign_latency_seconds -- Histogram **
  desc: The latency of egress IP assignment to ovn nb database

- name: egress_ips_unassign_latency_seconds -- Histogram **
  desc: The latency of egress IP unassignment from ovn nb database

- name: egress_ips_node_unreachable_total -- Counter
  desc: The total number of times assigned egress IP(s) were unreachable

- name: egress_ips_rebalance_total -- Counter
  desc: The total number of times assigned egress IP(s) needed to be moved to a different node

Add flags to explicitly enable the histogram metrics, since we only see value
in having them when scale testing egress ips. The flag introduced here is:
        `--metrics-enable-eip-scale`

Signed-off-by: Flavio Fernandes <flaviof@redhat.com>

### CODE REVIEW

File: go-controller/pkg/ovn/egressip.go [hunk 1]
```diff
@@ -1857,6 +1863,12 @@ type egressIPController struct {
 // (routing pod traffic to the egress node) and NAT objects on the egress node
 // (SNAT-ing to the egress IP).
 func (e *egressIPController) addPodEgressIPAssignment(egressIPName string, status egressipv1.EgressIPStatusItem, pod *kapi.Pod, podIPs []*net.IPNet) (err error) {
+	start := time.Now()
+	defer func() {
+		end := time.Since(start)
+		metrics.RecordEgressIPAssign(end)
```
Comment (kyrtapz) [inline go-controller/pkg/ovn/egressip.go:1869]: would it be ok to rename to `duration`?

File: go-controller/pkg/metrics/master.go [hunk 2]
```diff
@@ -189,6 +189,43 @@ var metricEgressIPCount = prometheus.NewGauge(prometheus.GaugeOpts{
...
+	Subsystem: MetricOvnkubeSubsystemMaster,
+	Name:      "egress_ips_node_unreachable_total",
+	Help:      "A metric that captures the number of times egress ip detected an unreachable node"},
+)
+
+var metricEgressIPRebalanceCount = prometheus.NewCounter(prometheus.CounterOpts{
+	Namespace: MetricOvnkubeNamespace,
+	Subsystem: MetricOvnkubeSubsystemMaster,
+	Name:      "egress_ips_rebalance_total",
+	Help:      "A metric that captures the number of times egress ip became invalid and had to be rebalanced"},
```
Comment (martinkennelly) [inline go-controller/pkg/metrics/master.go:226]: ip -> IP

File: go-controller/pkg/metrics/master.go [hunk 3]
```diff
@@ -189,6 +189,43 @@ var metricEgressIPCount = prometheus.NewGauge(prometheus.GaugeOpts{
 	Help:      "The number of defined egress IP addresses",
 })
 
+var metricEgressIPAssignLatency = prometheus.NewHistogram(prometheus.HistogramOpts{
```
Comment (martinkennelly) [inline go-controller/pkg/metrics/master.go:192]: Flavio, Can you help me understand why you want to expose a metric to record the configuration latency of egress ip? Trying to understand who would be interested in this latency and why.

File: go-controller/pkg/metrics/master.go [hunk 4]
```diff
@@ -189,6 +189,43 @@ var metricEgressIPCount = prometheus.NewGauge(prometheus.GaugeOpts{
...
+
+var metricEgressIPUnassignLatency = prometheus.NewHistogram(prometheus.HistogramOpts{
+	Namespace: MetricOvnkubeNamespace,
+	Subsystem: MetricOvnkubeSubsystemMaster,
+	Name:      "egress_ips_unassign_latency_seconds",
+	Help:      "The latency of egress ip unassignment from ovn nb database",
+	Buckets:   prometheus.ExponentialBuckets(.001, 2, 15),
+})
+
+var metricEgressIPNodeReacheableCount = prometheus.NewCounter(prometheus.CounterOpts{
```
Comment (martinkennelly) [inline go-controller/pkg/metrics/master.go:208]: why do we care if its reachable? Should we only care if its not reachable?

File: go-controller/pkg/metrics/master.go [hunk 5]
```diff
@@ -189,6 +189,43 @@ var metricEgressIPCount = prometheus.NewGauge(prometheus.GaugeOpts{
...
+	Namespace: MetricOvnkubeNamespace,
+	Subsystem: MetricOvnkubeSubsystemMaster,
+	Name:      "egress_ips_node_reachable_total",
+	Help:      "A metric that captures the number of times egress ip detected a reachable node"},
+)
+
+var metricEgressIPNodeUnreacheableCount = prometheus.NewCounter(prometheus.CounterOpts{
```
Comment (martinkennelly) [inline go-controller/pkg/metrics/master.go:215]: I like this metric because it will show us how often egressIPs are unreachable and may aid debugging for cluster admins.

File: go-controller/pkg/metrics/master.go (hunk 4 above)
Comment (martinkennelly) [reply go-controller/pkg/metrics/master.go:208]: We already have a metric where we know the number of assigned egress IPs so we have an idea already how many egress IPs are assigned. Do we need to know how many successful reach-ability attempts there were?

File: go-controller/pkg/metrics/master.go [hunk 6]
```diff
@@ -189,6 +189,43 @@ var metricEgressIPCount = prometheus.NewGauge(prometheus.GaugeOpts{
...
+	Namespace: MetricOvnkubeNamespace,
+	Subsystem: MetricOvnkubeSubsystemMaster,
+	Name:      "egress_ips_node_unreachable_total",
+	Help:      "A metric that captures the number of times egress ip detected an unreachable node"},
```
Comment (martinkennelly) [inline go-controller/pkg/metrics/master.go:219]: ip -> IP

What about:
`The total number of times assigned egress IP(s) were unreachable`

File: go-controller/pkg/metrics/master.go (hunk 2 above)
Comment (martinkennelly) [reply go-controller/pkg/metrics/master.go:226]: What about this.. I tried to dumb it down:
`The total number of times assigned egress IP(s) needed to be moved to a different node`

Am I right its only assigned egress IPs would need to be rebalanced?

File: go-controller/pkg/ovn/egressip.go [hunk 7]
```diff
@@ -1857,6 +1863,12 @@ type egressIPController struct {
 // (routing pod traffic to the egress node) and NAT objects on the egress node
 // (SNAT-ing to the egress IP).
 func (e *egressIPController) addPodEgressIPAssignment(egressIPName string, status egressipv1.EgressIPStatusItem, pod *kapi.Pod, podIPs []*net.IPNet) (err error) {
+	start := time.Now()
+	defer func() {
```
Comment (martinkennelly) [inline go-controller/pkg/ovn/egressip.go:1867]: I think this should be a level 5 log output. I may not understand why we need to record the latency.

File: go-controller/pkg/ovn/egressip.go [hunk 8]
```diff
@@ -1874,6 +1886,12 @@ func (e *egressIPController) addPodEgressIPAssignment(egressIPName string, statu
 // deletePodEgressIPAssignment deletes the OVN programmed egress IP
 // configuration mentioned for addPodEgressIPAssignment.
 func (e *egressIPController) deletePodEgressIPAssignment(egressIPName string, status egressipv1.EgressIPStatusItem, podIPs []*net.IPNet) error {
+	start := time.Now()
+	defer func() {
+		end := time.Since(start)
```
Comment (martinkennelly) [inline go-controller/pkg/ovn/egressip.go:1891]: Also, I think this should be level 5 log output. I maybe not understanding why you want to record the latency.

File: go-controller/pkg/ovn/egressip.go (hunk 1 above)
Comment (flavio-fernandes) [reply go-controller/pkg/ovn/egressip.go:1869]: sure thing! will do

File: go-controller/pkg/metrics/master.go (hunk 4 above)
Comment (flavio-fernandes) [reply go-controller/pkg/metrics/master.go:208]: good point. It may be overkill, so let's get rid of it and just keep metricEgressIPNodeUnreacheableCount

File: go-controller/pkg/metrics/master.go (hunk 3 above)
Comment (flavio-fernandes) [reply go-controller/pkg/metrics/master.go:192]: anserwerd offline; but in a nutshell, this was an idea for helping out with the scaling tests of egressip

File: go-controller/pkg/metrics/master.go (hunk 2 above)
Comment (flavio-fernandes) [reply go-controller/pkg/metrics/master.go:226]: right. good change!

File: go-controller/pkg/ovn/egressip.go (hunk 7 above)
Comment (flavio-fernandes) [reply go-controller/pkg/ovn/egressip.go:1867]: Talked some off offline. We will keep this as a metric, but will disable it by default, as it is only expected to be useful for egress scaling tests.

File: go-controller/pkg/ovn/egressip.go (hunk 8 above)
Comment (flavio-fernandes) [reply go-controller/pkg/ovn/egressip.go:1891]: We will keep this as a metric, but will disable it by default, as it is only expected to be useful for egress scaling tests.

File: go-controller/pkg/ovn/egressip.go (hunk 7 above)
Comment (flavio-fernandes) [reply go-controller/pkg/ovn/egressip.go:1868]: @martinkennelly please help with the flags here! I wonder if you think we could piggy back on `ovnkube_config_duration_enable` or create a new one

**Review by flavio-fernandes — COMMENTED** [review]
~~missing flag!~~

File: go-controller/pkg/ovn/egressip.go (hunk 7 above)
Comment (flavio-fernandes) [reply go-controller/pkg/ovn/egressip.go]: Disabled by default. To enable this metric, use newly added flag: `--metrics-enable-eip-scale`

File: go-controller/pkg/ovn/egressip.go (hunk 8 above)
Comment (flavio-fernandes) [reply go-controller/pkg/ovn/egressip.go]: Disabled by default. To enable this metric, use newly added flag: `--metrics-enable-eip-scale`

**Comment by flavio-fernandes** [comment]
@martinkennelly PTAL!

File: go-controller/pkg/metrics/master.go [hunk 9]
```diff
@@ -433,6 +469,26 @@ func RecordEgressIPCount(count float64) {
...
+func RecordEgressIPUnassign(duration time.Duration) {
+	metricEgressIPUnassignLatency.Observe(duration.Seconds())
+}
+
+// RecordEgressIPReachableNode records how many times EgressIP detected an unuseable node.
+func RecordEgressIPUnreachableNode() {
+	metricEgressIPNodeUnreacheableCount.Inc()
+}
+
+// RecordEgressIPReachableNode records how many times EgressIP detected an unuseable node.
```
Comment (martinkennelly) [inline go-controller/pkg/metrics/master.go:487]: correct the function description

**Review by martinkennelly — COMMENTED** [review]
/lgtm

Besides one tiny doc correction.
I know we agreed on slack to have a more generic flag for this but I think its fine to build up a few of these scale flags first and then later consolidate them.

**Comment by martinkennelly** [comment]
This isnt being enabled on CNO right?

File: go-controller/pkg/ovn/egressip.go [hunk 10]
```diff
@@ -1885,6 +1891,13 @@ type egressIPController struct {
 // (routing pod traffic to the egress node) and NAT objects on the egress node
 // (SNAT-ing to the egress IP).
 func (e *egressIPController) addPodEgressIPAssignment(egressIPName string, status egressipv1.EgressIPStatusItem, pod *kapi.Pod, podIPs []*net.IPNet) (err error) {
+	if config.Metrics.EnableEIPScaleMetrics {
+		start := time.Now()
+		defer func() {
+			duration := time.Since(start)
+			metrics.RecordEgressIPAssign(duration)
```
Comment (npinaeva) [inline go-controller/pkg/ovn/egressip.go:1898]: should be also recorded in case of error?

**Comment by flavio-fernandes** [comment]
> This isnt being enabled on CNO right?

Right @martinkennelly . The plan is to have QE enabling it when performing scaling tests. Not normally enabled.

File: go-controller/pkg/ovn/egressip.go (hunk 10 above)
Comment (flavio-fernandes) [reply go-controller/pkg/ovn/egressip.go:1898]: > should be also recorded in case of error?

I am hoping that we would have a more centralized location for counting failures, @npinaeva . In other words, the retry object logic would be my ideal place for detecting failures. So not part of the PR, I am hoping.

Comment (npinaeva) [reply go-controller/pkg/ovn/egressip.go:1898]: I mean since it is updated in defer, it will be updated regardless of the operation result. If you don't want to count error handling time, maybe add `err == nil` check before recording the metric?

Comment (flavio-fernandes) [reply go-controller/pkg/ovn/egressip.go:1898]: > I mean since it is updated in defer, it will be updated regardless of the operation result. If you don't want to count error handling time, maybe add `err == nil` check before recording the metric?

ah, sorry for missing that. Let me make that change.

Comment (flavio-fernandes) [reply go-controller/pkg/ovn/egressip.go:1938]: done. @npinaeva PTAL

Comment (npinaeva) [reply go-controller/pkg/ovn/egressip.go:1938]: Looks good now!
